View transitions 3.0 changes (#4320)
withastro/docs: src/content/docs/en/guides/view-transitions.mdx

Astro provides a `<ViewTransitions />` routing component that can be added to a single page's `<head>` to control page transitions as you navigate away to another page. It provides a lightweight client-side router that intercepts navigation and allows you to customize the transition between pages. Add this component to a reusable `.astro` component, such as a common head or layout, for animated page transitions across your entire site (SPA mode).

Astro's view transitions support is powered by the new [View Transitions](https://developer.chrome.com/docs/web-platform/view-transitions/) browser API and also includes:

- A few [built-in animations](#built-in-animation-directives), such as `fade`, `slide`, and `none`.
- Support for both forwards and backwards navigation animations.
- The ability to fully [customize all aspects of transition animation](#customizing-animations), and build your own animations.
- [Control over fallback behavior](#fallback-control) for browsers that do not yet support the View Transition APIs.

## Full site view transitions (SPA mode)

> Import and add the `<ViewTransitions />` component to your common `<head>` or shared layout component. Astro will create default page animations based on the similiarities between the old and new page, and will also provide fallback behavior for unsupported browsers.

Reviewer: Not added in this PR, but since we're already here...
```suggestion
Import and add the `<ViewTransitions />` component to your common `<head>` or shared layout component. Astro will create default page animations based on the similarities between the old and new page, and will also provide fallback behavior for unsupported browsers.
```

Add this component to a reusable `.astro` component, such as a common head or layout, for [animated page transitions across your entire site (SPA mode)](#full-site-view-transitions-spa-mode).

Astro's view transitions support is powered by the new [View Transitions](https://developer.chrome.com/docs/web-platform/view-transitions/) browser API and also includes:

- A few [built-in animation options](#built-in-animation-directives), such as `fade`, `slide`, and `none`.
- Support for both forwards and backwards navigation animations.
- The ability to fully [customize all aspects of transition animation](#customizing-animations), and build your own animations.
- The option to [prevent client-side navigation for non-page links](#preventing-client-side-navigation).
- [Control over fallback behavior](#fallback-control) for browsers that do not yet support the View Transition APIs.
- Automatic support for [`prefers-reduced-motion`](#prefers-reduced-motion).

:::note

> By default, every page will use regular, full-page, browser navigation. You must opt in to view transitions, and can use them either on a per-page basis or site-wide.

Reviewer: ```suggestion
By default, every page will use regular, full-page, browser navigation. You must opt in to view transitions and can use them either on a per-page basis or site-wide.
```

> Note that the `transition:name` can only be used once on each page. Set this manually when Astro can't infer a proper name itself, or for more fine control over matching elements.

Author: ```suggestion
Note that the provided `transition:name` value can only be used once on each page. Set this manually when Astro can't infer a proper name itself, or for more fine control over matching elements.
```

Reading this, it sounded like you could only use the directive once per page, but we mean that value can only be used once.

## Built-in Animation Directives

> Astro comes with a few built-in animations to override the default `fade` transition. Add the `transition:animate` directive to customize the behavior of your transitions.

Reviewer 2: ```suggestion
Astro comes with a few built-in animations to override the default `fade` transition. Add the `transition:animate` directive to individual elements to customize the behavior of specific transitions.
```

- `fade` (default): An opinionated crossfade animation, where the old content fades out and the new content fades in.

> - `initial`: Opt-out of Astro's opinionated crossfade animation and use the browser's default styling.

Reviewer: I guess since we're this as a verb it should be without the hyphen? Will let @sarah11918 make the final call here!
```suggestion
- `initial`: Opt out of Astro's opinionated crossfade animation and use the browser's default styling.
```

Reviewer 2: Yan knows his verbs!

- `none`: Disable the browser's default animations. This value can be used on a page's `<html>` element to disable the default fade.
- `slide`: An animation where the old content slides out to the left and new content slides in from the right. On backwards navigation, the animations are the opposite.

> The example below produces a slide animation for the body content while disabling the browser's fade animation for the rest of the page:

Reviewer 2: ```suggestion
Combine directives for full control over your page animation. Set a page default on the `<html>` element, and override on any individual elements as desired.

The example below produces a slide animation for the body content while disabling the browser's default fade animation for the rest of the page:
```

```astro
---
import { CommonHead } from '../components/CommonHead.astro';
---

<html transition:animate="none">
  <head>
    <CommonHead />
  </head>
  <body>
    <header>
      ...
    </header>

>     <main transition:animate="slide">

Reviewer 2: ```suggestion
    <!-- Override your page default on a single element -->
    <main transition:animate="slide">
```

```

:::note[known limitations]
The `initial` browser animation is not simulated by Astro. So any element using this animation will not currently be animated.
:::

## Navigation steps

When using the `<ViewTransitions />` router, the following steps are performed as part of navigation:

1. The user triggers navigation by either:
    - Click a `<a>` to another page in your site.
    - Clicking the back button.
    - Clicking the forward button.
2. The router starts fetching the next page.
3. The router adds the `data-astro-transition` attribute to the HTML element with a value of `'forward'` or `'back'` depending on whether it is back navigation or not.

> 4. The router calls `document.startViewTransition`. This triggers the browsers own [view transition process](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API#the_view_transition_process). Importantly the browser screenshots the current state of the page.

Reviewer: ```suggestion
4. The router calls `document.startViewTransition`. This triggers the browser's own [view transition process](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API#the_view_transition_process). Importantly the browser screenshots the current state of the page.
```

<header transition:animate={myFade}> ... </header>
```

## Preventing client-side navigation

Once you have enabled client-side routing on a page by adding the `<ViewTransitions />` component, every anchor on the page that links to another page on your site will be navigated via client-side routing. There are some cases where you might want to not navigate via CSR. One example is if the link is to any non-page, such as a PDF within your `public/` folder. Or an API route that produces an image.

> In these cases you can opt-out of client-side routing on a per-link basis using the `data-astro-reload` attribute like so:

Reviewer: ```suggestion
In these cases, you can opt out of client-side routing on a per-link basis using the `data-astro-reload` attribute like so:
```

```

:::note[known limitations]
The `initial` browser animation is not simulated by Astro. So any element using this animation will not currently be animated.
:::

## Client-side navigation process

When using the `<ViewTransitions />` router, the following steps occur to produce Astro's client-side navigation:

1. A visitor to your site triggers navigation by any of the following actions:

>     - Clicking an `<a>` tag linking internally to another page in your site.

Reviewer: I guess it's "on" here?
```suggestion
    - Clicking an `<a>` tag linking internally to another page on your site.
```

  // Runs on initial navigation
	setDarkMode();
  // Runs on view transitions navigation
	document.addEventListener('astro:after-swap', setDarkMode);
</script>
```

## `prefers-reduced-motion`

Astro's `<ViewTransitions />` component includes a CSS media query that disables *all* view transition animations, including fallback animation, whenever the [`prefer-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) setting is detected. Instead, the browser will simply swap the DOM elements without an animation.

## Preventing client-side navigation

> Once you have enabled client-side routing on a page by adding the `<ViewTransitions />` component, every anchor on the page that links to another page on your site will be navigated via client-side routing. There are some cases where you might want to not navigate via CSR. One example is if it he link is to any non-page, such as a PDF within your `public/` folder. Or an API route that produces an image.

Reviewer 3: ```suggestion
Once you have enabled client-side routing on a page by adding the `<ViewTransitions />` component, every anchor on the page that links to another page on your site will be navigated via client-side routing. There are some cases where you might want to not navigate via CSR. One example is if the link is to any non-page, such as a PDF within your `public/` folder. Or an API route that produces an image.
```

</script>
```

## `prefers-reduced-motion`

Astro's `<ViewTransitions />` component includes a CSS media query that disables *all* view transition animations, including fallback animation, whenever the [`prefer-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) setting is detected. Instead, the browser will simply swap the DOM elements without an animation.

## Preventing client-side navigation

Once you have enabled client-side routing on a page by adding the `<ViewTransitions />` component, every anchor on the page that links to another page on your site will be navigated via client-side routing. There are some cases where you might want to not navigate via CSR. One example is if the link is to any non-page, such as a PDF within your `public/` folder. Or an API route that produces an image.

In these cases you can opt-out of client-side routing on a per-link basis using the `data-astro-reload` attribute like so:

```html
<a href="/quarterly-earnings.pdf" data-astro-reload>
```

These links will be ignored by the router and a full page navigation will occur.

## Upgrade to v3.0 from v2.x

> View transitions are no longer behind an experimental flag in Astro v3.0.

Reviewer 3: Added an upgrade guide to this page as discussed!

Reviewer 3: Probably needs to mention the changed event names as well @matthewp

```js title="astro.config.mjs" del={4-6}
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {
   viewTransitions: true
  }
});
```

#### Update `transition:animate` directives

**Changed:** The `transition:animate` value `morph` has been renamed to `initial`. Also, this is no longer the default animation. If no `transition:animate` directive is specified, your animations will now default to `fade`.

1. Rename any `morph` animations to `intial`.
    ```astro title="src/components/MyComponent.astro" del="morph" ins="initial"
    <div transition:name="name" transition:animate="morph initial" />
    ```
2. To keep any animations that were previously using `morph` by default, explicitly add `transition:animate="initial"`

    ```astro title="src/components/MyComponent.astro" ins='transition:animate="initial"'
    <div transition:name="name" transition:animate="initial" />
    ```
3. You can safely remove any animations explicitly set to `fade`. This is now the default behavior:

    ```astro title="src/components/MyComponent.astro" del="transition:animate=\"fade\""
    <div transition:name="name" transition:animate="fade" />
    ```

> **Added:** Astro also supports a new `transition:animate` value, `none`. This value can be used on a page's `<html>` element to disable animated full-page transitions on an entire page. This will only override **default animation behavior** on page elements without an animation directive. You can still set animations on individual elements, and these specific animations will be occur.

Reviewer: "be"
```suggestion
**Added:** Astro also supports a new `transition:animate` value, `none`. This value can be used on a page's `<html>` element to disable animated full-page transitions on an entire page. This will only override **default animation behavior** on page elements without an animation directive. You can still set animations on individual elements, and these specific animations will occur.
```

4. You may now disable all default transitions on an individual page, animating only elements that explicitly use a `transition:animate` directive: 

    ```astro ins="transition:animate=\"none\""
    <html transition:animate="none">
      <head></head>
      <body>
        <h1>Hello world!</h1>
      </body>
    </html>
    ```

#### Update event names

> The event `astro:load` has been renamed to `astro:page-load`. Rename all occurances in your project.
> 
> ```astro title="src/components/MyComponent.astro" del="astro:load" ins="astro:page-load"
> <script>
> document.addEventListener('astro:load astro:page-load', runSetupLogic);
> </script>
> ```
> 
> The event `astro:beforeload` has been renamed to `astro:after-swap`. Rename all occurances in your project.

Reviewer: `occurances` -> `occurrences`
````suggestion
The event `astro:load` has been renamed to `astro:page-load`. Rename all occurrences in your project.

```astro title="src/components/MyComponent.astro" del="astro:load" ins="astro:page-load"
<script>
document.addEventListener('astro:load astro:page-load', runSetupLogic);
</script>
```

The event `astro:beforeload` has been renamed to `astro:after-swap`. Rename all occurrences in your project.
````
